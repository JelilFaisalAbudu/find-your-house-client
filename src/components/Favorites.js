import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import houseImage from '../images/scott-webb-1ddol8rgUH8-unsplash.jpg';
import userFavorites from '../redux/actions/favorite';
import getHouses from '../redux/actions/houses';
import { spyNoScroll } from '../helpers/dom';

const FavoritesHouses = () => {
  const dispatch = useDispatch();
  const { user } = useSelector(state => state.auth);
  const { favoriteHousesIds } = useSelector(state => state.favorites);
  const { houses } = useSelector(state => state.houses);

  const getUserFavorites = userId => {
    dispatch(userFavorites.getUserFavorites(userId));
  };
  const getFavoritesHouses = (houses, ids) => houses
    .filter(house => ids.includes(house.id));

  const handleRemoveFavorite = (userId = user.id, houseId) => {
    dispatch(userFavorites.removeUserFavorite(userId, houseId));
    dispatch(userFavorites.doRemoveFavorite(houseId));
  };

  useEffect(() => {
    dispatch(getHouses());
    spyNoScroll();
  }, []);

  useEffect(() => {
    if (user) {
      getUserFavorites(user.id);
    }
  }, []);

  const content = (

    <div className="container-fluid">
      <div className="favorite-houses row g-2">
        {getFavoritesHouses(houses, favoriteHousesIds).length
          ? (getFavoritesHouses(houses, favoriteHousesIds).map(house => (
            <div key={house.id} className="house col-12 col-lg-6 col-xlg-4 text-white position-relative">
              <button
                type="button"
                aria-label="Remove from favorite list"
                className="btn btn-unlike"
                data-toggle="tooltip"
                data-placement="top"
                title="remove from favorite list"
                onClick={() => handleRemoveFavorite(user.id, house.id)}
              >
                <span className="material-icons">
                  favorite
                </span>
              </button>
              <div className="house-info">
                <h2 className="display-5">{house.name}</h2>
                <h2 className="display-5">{house.category}</h2>
                <p className="lead">And an even wittier subheading.</p>
              </div>
              <img
                className="img-fluid"
                src={houseImage}
                alt={house.name}
              />
            </div>

          )))
          : (<div className="col-12">You have no favorite house...</div>)}
      </div>
    </div>

  );

  return content;
};

export default FavoritesHouses;
