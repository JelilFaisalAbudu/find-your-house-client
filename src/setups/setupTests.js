import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

Enzyme.configure({ adapter: new Adapter() });
