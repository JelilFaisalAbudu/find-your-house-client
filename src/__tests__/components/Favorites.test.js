import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../setups/setupTests';
import Favorites from '../../components/Favorites';
import store from '../../redux/store';

const setup = () => mount(
  <Provider store={store}>
    <Favorites />
  </Provider>,
);

describe('Favorites', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  test('renders without crashing', () => {
    // console.log(toJson(wrapper));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('', () => {
    const user = {
      id: 4,
      name: 'John Doe',
      auth_token: 'doiifhfieiffhghwnc4429hBD',
    };
    const houses = [
      {
        id: 1,
        name: 'Golden Lu',
        category: 'Glass House',
        description: 'Lorem Ipsum',
        imageUrl: 'http://example.com/golden-lu.jpg',
      },
      {
        id: 2,
        name: 'Golden Lu',
        category: 'Glass House',
        description: 'Lorem Ipsum',
        imageUrl: 'http://example.com/golden-lu.jpg',
      },
      {
        id: 3,
        name: 'Golden Lu',
        category: 'Glass House',
        description: 'Lorem Ipsum',
        imageUrl: 'http://example.com/golden-lu.jpg',
      },
    ];

    const loginUserAction = user => ({
      type: 'LOGIN_SUCCESS',
      payload: { user },
    });

    const addHouses = houses => ({
      type: 'GET_HOUSES',
      payload: { houses },
    });

    const addFavoriteAction = house => ({
      type: 'ADD_FAVORITE',
      payload: { house },
    });

    beforeAll(() => {
      store.dispatch(addHouses(houses));
      store.dispatch(loginUserAction(user));
    });

    test('renders the Favorites page correctly but with no favorite house', async () => {
      const favoriteHouses = await wrapper
        .find('div.houses-container');
      expect(favoriteHouses.text()).toMatch(/You have no favorite house.../i);
    });

    test('should add a new favorite house for the user', () => {
      const { auth, houses, favorites } = store.getState();
      const newFavorite = {
        user_id: auth.user.id,
        house_id: houses.houses[2],
      };
      store.dispatch(addFavoriteAction(newFavorite));

      const newFavoritesState = store.getState().favorites;

      expect(newFavoritesState.favoriteHouses.length)
        .toBeGreaterThan(favorites.favoriteHouses.length);
    });
  });
});
