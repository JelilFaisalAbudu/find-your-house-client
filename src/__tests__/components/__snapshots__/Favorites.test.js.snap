// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <FavoritesHouses>
    <div
      className="container-fluid"
    >
      <div
        className="favorite-houses row g-2"
      >
        <div
          className="col-12"
        >
          You have no favorite house...
        </div>
      </div>
    </div>
  </FavoritesHouses>
</Provider>
`;
